src/AzurePipelines/PipelineCachingCacheClient.cs

```diff
@@ -219,64 +222,177 @@ internal sealed class PipelineCachingCacheClient : CacheClient
+                }
 
-                // 2. Link out unique content to the temp folder
+                // create the manifest and add extras to local cache
+                Manifest manifest;
+                {
+                    var items = new List<ManifestItem>(outputs.Count + extras.Count);
 
-                Dictionary<ContentHash, string> tempFilesPerHash = outputs.Values.Distinct().ToDictionary(
-                    hash => hash,
-                    hash =>
+                    // put extras in local cache to simplify the code below
+                    foreach (KeyValuePair<string, FileInfo> extra in extras)
                     {
-                        string tempFilePath = Path.Combine(TempFolder, Guid.NewGuid().ToString("N") + ".tmp");
-                        tempFilePaths.Add(tempFilePath);
-                        return tempFilePath;
-                    });
+                        DedupNode node = await ChunkerHelper.CreateFromFileAsync(FileSystem.Instance, extra.Key, cancellationToken, configureAwait: false);
+                        DedupIdentifier dedupId = node.GetDedupIdentifier();
+                        dedupToSize[dedupId] = extra.Value.Length;
+                        dedupToHash[dedupId] = node.ToContentHash(_hashType);
+                        await LocalCacheSession.PutFileAsync(context, node.ToContentHash(_hashType), new AbsolutePath(extra.Value.FullName), FileRealizationMode.Any, cancellationToken);
+                        items.Add(new ManifestItem(extra.Key, new DedupInfo(dedupId.ValueString, node.TransitiveContentBytes)));
+                    }
+
+                    foreach (KeyValuePair<string, ContentHash> output in outputs)
+                    {
+                        string relativePath = output.Key.MakePathRelativeTo(RepoRoot)!.Replace("\\", "/", StringComparison.Ordinal);
+                        DedupIdentifier dedupId = output.Value.ToBlobIdentifier().ToDedupIdentifier();
+                        items.Add(new ManifestItem(
+                            relativePath,
+                            new DedupInfo(dedupId.ValueString, (ulong)dedupToSize[dedupId])));
+                    }
+                    items.Sort((i1, i2) => StringComparer.Ordinal.Compare(i1.Path, i2.Path));
 
-                List<ContentHashWithPath> tempFiles = tempFilesPerHash
-                    .Select(kvp => new ContentHashWithPath(kvp.Key, new AbsolutePath(kvp.Value)))
-                    .ToList();
+                    manifest = new Manifest(items);
+                }
 
-                Dictionary<string, PlaceFileResult> placeResults = await TryPlaceFilesFromCacheAsync(context, tempFiles, cancellationToken);
-                foreach (PlaceFileResult placeResult in placeResults.Values)
+                // Store the manifest in local cache to simplify the code below
+                using MemoryStream manifestStream = new(JsonSerializer.Serialize(manifest).GetUTF8Bytes());
```

Comment:
I believe you can serialize directly to a stream, which should cut out the intermediate string.